servershr/ServerDispatchPhase.c

```diff
@@ -634,20 +634,16 @@ static void dispatch(int i)
     }
     else
     {
+      actions[i].dispatched = 1;
       UNLOCK_ACTION(i, d_w);
       status = ServerDispatchAction(
           0, Server(server, actions[i].server), table->tree, table->shot,
           actions[i].nid, action_done, (void *)(intptr_t)i, &actions[i].status,
           &actions[i].lock, &actions[i].netid, before);
-      if (STATUS_OK)
-      {
-        WRLOCK_ACTION(i, d_w);
-        actions[i].dispatched = 1;
-        UNLOCK_ACTION(i, d);
-      }
-      else
+      if (STATUS_NOT_OK)
```

Comment:
Note that `ServerDispatchAction()` can return `SsINTERNAL` which is not an MDSplus status code.   The `SsINTERNAL` evaluates to `-1`.   And that value has the low-order bit set, thus is erroneously treated as `STATUS_OK`.   And thus this `if` statement likely fails.

Global question . . . 

There are several instances in the entire source code that fail because the check for `SsINTERNAL` is missing.  So, I'm wondering if it is safe to redefine `SsINTERNAL` to 0xFFFF'FFFE` so the low-order bit is zero, and thus compatible with the `STATUS_OK` and `STATUS_NOT_OK` macros?   

That would be a massive change because it impacts the entire code base (all APIs, all languages).   So, to be conservative, I've instead been fixing `if` statements to also check for `SsINTERNAL`.
- e.g., `if ((status == SsINTERNAL) || STATUS_NOT_OK)`

However, it sure would be nice if we could just redefine `SsINTERNAL` without breaking anything, and be rid of this headache.

Reply:
the usage of SsInternal if not an error but used as an internal return value tgat should never surface. code that acts as mdsplus funtion but does not return a proper MDSplus status code is considered malformed. There is no use to assign a different code to SsInternal other that specifically treating tge most common malformed return value for error. on the otger hand returning 0 for success will result in the same missinterpretation.

Reply:
Makes sense.

Another possible global change would be to redefine the `IS_OK`, etc.  macros to add the check for `SsINTERNAL`.  (Also probably too risky a change.)   

Regardless, there are several instances where low-level routines return `SsINTERNAL` and the calling routines don't check for it, and thus `SsINTERNAL` can ripple up the call stack and eventually cause problems.

Reply:
I have opened Issue #2741 regarding the problems with non-MDSplus error codes being ignored by calling routines.   That is a global issue that we will not solve as part of the fix for Issue #2731.

At present, we will limit the scope to just fixing error handling associated with Issue #2731.